components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs

```diff
@@ -31,20 +31,70 @@ public class WriterGroupMessageTrigger : IMessageTrigger, IDisposable {
             Subscription?.NumberOfConnectionRetries ?? 0;
 
         /// <inheritdoc/>
-        public int ValueChangesCount { get; private set; } = 0;
+        public ulong ValueChangesCountLastMinute {
+            get => CalculateSumForRingBuffer(_valueChangesBuffer);
+            private set => IncreaseRingBuffer(_valueChangesBuffer, ref _lastPointerValueChanges, _bucketWidth, value);
+        }
+
+        /// <inheritdoc/>
+        public ulong ValueChangesCount {
+            get { return _valueChangesCount; }
+            private set  {
+                var difference = value - _valueChangesCount;
+                _valueChangesCount = value;
+                ValueChangesCountLastMinute = difference;
+            }
+        }
 
         /// <inheritdoc/>
-        public int DataChangesCount { get; private set; } = 0;
+        public ulong DataChangesCountLastMinute {
+            get => CalculateSumForRingBuffer(_dataChangesBuffer);
+            private set => IncreaseRingBuffer(_dataChangesBuffer, ref _lastPointerDataChanges, _bucketWidth, value);
+        }
+
+        /// <inheritdoc/>
+        public ulong DataChangesCount {
+            get {
+                return _dataChangesCount;
+            }
+            private set {
+                var difference = value - _dataChangesCount;
+                _dataChangesCount = value;
+                DataChangesCountLastMinute = difference;
+            }
+        }
+
+        /// <summary>
+        /// Iterates the array and add up all values
+        /// </summary>
+        private static ulong CalculateSumForRingBuffer(ulong[] array) {
+            ulong sum = 0;
+            for(int index = 0; index< array.Length; index++) {
+                sum += array[index];
+            }
+            return sum;
+        }
+
+        /// <summary>
+        /// Helper function to distribute values over array based on time
+        /// </summary>
+        private void IncreaseRingBuffer(ulong[] array, ref int lastPointer, int bucketWidth, ulong difference) {
+            var indexPointer = DateTime.UtcNow.Second % bucketWidth;
+            while (lastPointer != indexPointer && (lastPointer + 1) < array.Length) {
+                lastPointer++;
+                array[lastPointer] = 0;
+            }
+
```

Comment:
the wrap around case seems not quite right:

while (lastPointer != indexPointer)
 {
                lastPointer = lastPointer++ % bucketWidth;
                array[lastPointer] = 0;
            }

Reply:
changed